File: src/eggpool/health/health_manager.py
"""Health management for accounts and models."""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any

from eggpool.health.circuit_breaker import CircuitBreaker


class FailureCategory(StrEnum):
    """Normalized failure categories used across health and runtime state."""

    AUTHENTICATION_FAILED = "authentication_failed"
    QUOTA_EXHAUSTED = "quota_exhausted"
    RATE_LIMITED = "rate_limited"
    MODEL_UNAVAILABLE = "model_unavailable"
    CONNECT_TIMEOUT = "connect_timeout"
    CONNECTION_FAILURE = "connection_failure"
    UPSTREAM_SERVER_ERROR = "upstream_server_error"
    PROTOCOL_ERROR = "protocol_error"
    CONTEXT_LIMIT_EXCEEDED = "context_limit_exceeded"
    UNKNOWN = "unknown"


def classify_failure_category(
    error_class: str | None,
    status_code: int | None = None,
) -> FailureCategory:
    """Map an error class string or status code to a normalized failure category.

    Real upstream HTTP 402 (Payment Required) responses map to
    QUOTA_EXHAUSTED regardless of error class, because providers often
    return generic 402 bodies without a structured error class. The
    quota-exhausted substring check is intentionally permissive to
    accept vendor-specific spellings (``quotaexhausted`` or
    ``quota_exhausted``).

    Phase 6 explicit handling:

    * HTTP 408 (request timeout) maps to ``CONNECT_TIMEOUT`` so a
      timed-out upstream request is treated as transport pressure
      rather than a client mistake.
    * HTTP 409 and 422 are intentionally mapped to ``UNKNOWN``; they
      are provider-specific and must not trigger account suppression
      unless the error class explicitly matches a known category.
    """
    if error_class is None and status_code is None:
        return FailureCategory.UNKNOWN
    if status_code == 402:
        return FailureCategory.QUOTA_EXHAUSTED
    if status_code == 408:
        return FailureCategory.CONNECT_TIMEOUT
    if status_code in (409, 422):
        # Provider-specific; do not blindly suppress account health
        # unless the body / error class explicitly identifies a
        # category the caller already knows about.
        return FailureCategory.UNKNOWN
    if error_class is None:
        # status_code already handled above; any other code without
        # an error class falls through to the generic catch-all below.
        if status_code is not None and 500 <= status_code < 600:
            return FailureCategory.UPSTREAM_SERVER_ERROR
        return FailureCategory.UNKNOWN
    ec = error_class.lower()
    if "contextlimitexceeded" in ec or "context_limit_exceeded" in ec:
        return FailureCategory.CONTEXT_LIMIT_EXCEEDED
    if "auth" in ec:
        return FailureCategory.AUTHENTICATION_FAILED
    if "quotaexhausted" in ec or "quota_exhausted" in ec:
        return FailureCategory.QUOTA_EXHAUSTED
    if "ratelimit" in ec or "rate_limit" in ec or status_code == 429:
        return FailureCategory.RATE_LIMITED
    if "modelunavailable" in ec or "model_not_found" in ec:
        return FailureCategory.MODEL_UNAVAILABLE
    if "connecttimeout" in ec or "connect_timeout" in ec:
        return FailureCategory.CONNECT_TIMEOUT
    conn_terms = (
        "connectionfailure",
        "connection_failure",
        "connectionerror",
        "connecterror",
    )
    if any(s in ec for s in conn_terms):
        return FailureCategory.CONNECTION_FAILURE
    if "timeout" in ec:
        return FailureCategory.CONNECT_TIMEOUT
    if "temporary" in ec or "transient" in ec:
        if status_code is not None and 500 <= status_code < 600:
            return FailureCategory.UPSTREAM_SERVER_ERROR
        return FailureCategory.UNKNOWN
    if status_code is not None and 500 <= status_code < 600:
        return FailureCategory.UPSTREAM_SERVER_ERROR
    return FailureCategory.UNKNOWN


@dataclass(slots=True)
class AccountHealth:
    """Health state for an account."""

    account_name: str
    is_healthy: bool = True
    last_check: float = field(default_factory=time.time)
    consecutive_failures: int = 0
    circuit_breaker: CircuitBreaker = field(default_factory=CircuitBreaker)
    # model_id -> disabled_until timestamp (``None`` means disabled
    # indefinitely, matching the account-level ``disabled_until``
    # convention).
    disabled_models: dict[str, float | None] = field(
        default_factory=dict[str, float | None]
    )
    disabled_until: float | None = None
    disabled_reason: str = ""
    cooldown_until: float = 0.0
    health_state: str = "healthy"

    def is_disabled(self, current_time: float | None = None) -> bool:
        """Check if account is currently disabled."""
        if current_time is None:
            current_time = time.time()

        disabled = (
            self.disabled_until is not None and current_time < self.disabled_until
        )
        cooled = self.cooldown_until > 0 and current_time < self.cooldown_until
        return disabled or cooled

    def is_model_disabled(
        self, model_id: str, current_time: float | None = None
    ) -> bool:
        """Check if a model is disabled for this account."""
        if self.is_disabled(current_time):
            return True
        if model_id not in self.disabled_models:
            return False
        until = self.disabled_models[model_id]
        if until is None:
            return True
        now = current_time if current_time is not None else time.time()
        if now >= until:
            self.disabled_models.pop(model_id, None)
            return False
        return True


@dataclass(slots=True)
class HealthManager:
    """Manages health state for all accounts."""

    _accounts: dict[str, AccountHealth] = field(
        default_factory=dict[str, AccountHealth]
    )

    def get_account_health(self, account_name: str) -> AccountHealth:
        """Get or create account health."""
        if account_name not in self._accounts:
            self._accounts[account_name] = AccountHealth(account_name=account_name)
        return self._accounts[account_name]

    def record_success(self, account_name: str, model_id: str | None = None) -> None:
        """Record a successful request."""
        health = self.get_account_health(account_name)
        health.consecutive_failures = 0
        health.last_check = time.time()
        # An in-flight request may succeed after an operator disables its
        # account. Do not let that completion undo an explicit disable.
        if (
            health.health_state != "authentication_failed"
            and not health.disabled_reason
        ):
            health.is_healthy = True
            health.health_state = "healthy"
            health.cooldown_until = 0.0
        health.circuit_breaker.record_success()

    def record_failure(
        self,
        account_name: str,
        model_id: str | None = None,
        reason: str | None = None,
    ) -> None:
        """Record a failed request."""
        health = self.get_account_health(account_name)
        health.consecutive_failures += 1
        health.last_check = time.time()
        health.circuit_breaker.record_failure()
        if reason == "authentication_failed":
            health.health_state = "authentication_failed"
            health.is_healthy = False
        elif reason in ("rate_limited", "quota_exhausted"):
            health.health_state = reason
            health.is_healthy = False

    def record_quota_exhausted(
        self,
        account_name: str,
        cooldown_seconds: float,
    ) -> None:
        """Place account into a bounded quota-exhausted cooldown."""
        health = self.get_account_health(account_name)
        health.health_state = "quota_exhausted"
        health.cooldown_until = time.time() + cooldown_seconds
        health.is_healthy = False

    def record_rate_limit(self, account_name: str, retry_after_seconds: float) -> None:
        """Record a rate limit with explicit cooldown."""
        health = self.get_account_health(account_name)
        health.cooldown_until = time.time() + max(0.0, retry_after_seconds)
        health.health_state = "rate_limited"
        health.is_healthy = False

    def record_failure_with_policy(
        self,
        account_name: str,
        reason: str,
        *,
        retry_after: float | None = None,
    ) -> float | None:
        """Apply a reason-specific backoff policy to the account.

        Routes through the dedicated :mod:`eggpool.health.backoff`
        layer so the policy table is testable and reviewable in
        isolation. Returns the cooldown duration in seconds when one
        was applied, ``None`` otherwise (terminal or no-op reasons).

        Parameters
        ----------
        account_name:
            Target account. The method is a no-op for unknown
            accounts unless the call later creates one via the health
            manager's lookup.
        reason:
            A :class:`BackoffReason` value or string. Unknown reasons
            fall through to :meth:`record_failure` with the same
            ``reason``.
        retry_after:
            Optional upstream ``Retry-After`` value (seconds). Honored
            for ``rate_limited`` and ``quota_exhausted``.

        Notes
        -----
        ``model_unavailable`` is handled by calling
        :meth:`disable_model` for the account-wide scope only. The
        per-model key is intentionally not threaded through this
        method; the coordinator must invoke :meth:`disable_model`
        directly with the specific ``model_id`` so the catalog cache
        and health manager stay in sync.
        """
        # Local import keeps the health manager usable in environments
        # where the backoff module is intentionally mocked out.
        from eggpool.health.backoff import (
            BackoffPolicy,
            compute_backoff_seconds,
            get_backoff_policy,
        )

        if reason == "authentication_failed":
            self.record_failure(account_name, reason="authentication_failed")
            return None

        if reason == "context_limit_exceeded":
            # No account-level suppression for context-limit errors.
            return None

        policy: BackoffPolicy | None = get_backoff_policy(reason)
        if policy is None:
            # Unknown reason: fall back to the legacy record_failure
            # path so existing behavior is preserved.
            self.record_failure(account_name, reason=reason)
            return None

        if policy.base_delay <= 0 or policy.cap <= 0:
            # Terminal (auth) or zero-policy reasons already routed
            # above; this branch handles ``model_unavailable`` whose
            # policy exists but uses an empty base to signal "no
            # exponential backoff".
            if reason == "model_unavailable":
                # Per-model disable lives outside this method because
                # it requires the model id; the coordinator calls
                # ``disable_model`` directly.
                return None
            return None

        health = self.get_account_health(account_name)
        delay = compute_backoff_seconds(
            reason,
            consecutive_failures=health.consecutive_failures + 1,
            retry_after=retry_after,
            jitter=True,
        )
        if delay is None:
            return None

        if reason == "quota_exhausted":
            self.record_quota_exhausted(account_name, delay)
            return delay
        if reason == "rate_limited":
            self.record_rate_limit(account_name, delay)
            return delay

        # Generic transient category: bounded cooldown with the
        # generic "cooldown" health state so the runtime view stays
        # consistent with ``AccountRuntimeState``.
        health.cooldown_until = time.time() + delay
        health.health_state = "cooldown"
        health.is_healthy = False
        health.consecutive_failures += 1
        health.last_check = time.time()
        return delay

    def disable_account(
        self,
        account_name: str,
        reason: str,
        duration_seconds: float | None = None,
    ) -> None:
        """Disable an account."""
        health = self.get_account_health(account_name)
        health.is_healthy = False
        health.disabled_reason = reason
        health.disabled_until = (
            None
            if duration_seconds is None
            else time.time() + max(0.0, duration_seconds)
        )

    def enable_account(self, account_name: str) -> None:
        """Enable an account."""
        health = self.get_account_health(account_name)
        health.is_healthy = True
        health.health_state = "healthy"
        health.disabled_until = None
        health.disabled_reason = ""
        health.cooldown_until = 0.0
        health.circuit_breaker.reset()

    def disable_model(
        self,
        account_name: str,
        model_id: str,
        duration_seconds: float | None = None,
    ) -> None:
        """Disable a model for an account.

        Mirrors :meth:`disable_account`: when ``duration_seconds`` is
        provided, the disable expires after that interval. ``None``
        disables the model indefinitely until :meth:`enable_model` is
        called.
        """
        health = self.get_account_health(account_name)
        health.disabled_models[model_id] = (
            None
            if duration_seconds is None
            else time.time() + max(0.0, duration_seconds)
        )

    def enable_model(self, account_name: str, model_id: str) -> None:
        """Enable a model for an account."""
        health = self.get_account_health(account_name)
        health.disabled_models.pop(model_id, None)

    def prune_disabled_models(
        self,
        account_name: str,
        advertised_models: set[str],
    ) -> int:
        """Drop ``disabled_models`` entries no longer advertised upstream.

        ``disabled_models`` keys are ``model_id`` strings. An entry whose
        model has been withdrawn upstream is dead weight — the model is
        no longer reachable through this account, so the disable is
        moot. Returns the number of entries removed (intended for log
        diagnostics). Unknown accounts and empty ``disabled_models``
        are no-ops.
        """
        health = self._accounts.get(account_name)
        if health is None or not health.disabled_models:
            return 0
        stale = [mid for mid in health.disabled_models if mid not in advertised_models]
        for mid in stale:
            del health.disabled_models[mid]
        return len(stale)

    def _refresh_transient_state(self, health: AccountHealth) -> None:
        """Restore transient health states after cooldown expiration."""
        now = time.time()
        if (
            health.cooldown_until > 0
            and now >= health.cooldown_until
            and health.health_state in ("quota_exhausted", "rate_limited", "cooldown")
        ):
            health.health_state = "healthy"
            health.is_healthy = True
            health.cooldown_until = 0
        # Clear expired timed disable, but do not override terminal states
        # such as authentication_failed which require explicit enable.
        if (
            health.disabled_until is not None
            and now >= health.disabled_until
            and health.health_state != "authentication_failed"
        ):
            health.disabled_until = None
            health.disabled_reason = ""
            health.is_healthy = True

    def is_account_healthy(self, account_name: str) -> bool:
        """Check if an account is healthy."""
        health = self.get_account_health(account_name)
        self._refresh_transient_state(health)
        return health.is_healthy and not health.is_disabled()

    def is_model_healthy(self, account_name: str, model_id: str) -> bool:
        """Check if a model is healthy for an account.

        Uses the non-mutating :meth:`CircuitBreaker.can_request` so
        readiness probes and candidate enumeration never consume the
        half-open probe slot.
        """
        if not self.is_account_healthy(account_name):
            return False
        health = self.get_account_health(account_name)
        return (
            not health.is_disabled()
            and not health.is_model_disabled(model_id)
            and health.circuit_breaker.can_request()
        )

    def try_acquire_request(self, account_name: str, model_id: str) -> bool:
        """Attempt to acquire a circuit-breaker probe slot for dispatch.

        Calls the mutating :meth:`CircuitBreaker.allow_request` which
        transitions OPEN → HALF_OPEN and sets the in-flight flag when
        appropriate.  Returns ``True`` when the request may proceed,
        ``False`` when the circuit breaker rejects it (account should
        be excluded from this routing round).
        """
        if not self.is_account_healthy(account_name):
            return False
        health = self.get_account_health(account_name)
        if health.is_disabled() or health.is_model_disabled(model_id):
            return False
        return health.circuit_breaker.allow_request()

    def release_request(self, account_name: str) -> None:
        """Release the circuit-breaker half-open probe slot for an account.

        Call this when a request that acquired a probe slot terminates
        through a path that does not call :meth:`record_success` or
        :meth:`record_failure` (client cancellation, client error,
        rate-limit/quota cooldown, model disabled).
        """
        health = self.get_account_health(account_name)
        health.circuit_breaker.release_probe()

    def get_healthy_accounts(self, account_names: list[str]) -> list[str]:
        """Get list of healthy accounts."""
        return [name for name in account_names if self.is_account_healthy(name)]

    def get_health_stats(self, account_name: str) -> dict[str, Any]:
        """Get health statistics for an account."""
        health = self.get_account_health(account_name)
        self._refresh_transient_state(health)
        for model_id in list(health.disabled_models):
            health.is_model_disabled(model_id)
        return {
            "account_name": health.account_name,
            "is_healthy": health.is_healthy,
            "last_check": health.last_check,
            "consecutive_failures": health.consecutive_failures,
            "circuit_breaker": health.circuit_breaker.get_stats(),
            "disabled_models": list(health.disabled_models),
            "disabled_until": health.disabled_until or 0,
            "disabled_reason": health.disabled_reason,
        }

File: rust/src/health/health_manager.rs
//! Account health state and read-only versus mutating routing gates.

use std::{
    collections::{BTreeMap, BTreeSet},
    sync::{Arc, Mutex},
};

use thiserror::Error;

use super::{
    AccountBackoffRecord, BackoffReason, CircuitBreaker, CircuitStats,
    MAX_NONTERMINAL_BACKOFF_SECONDS,
};

#[derive(Debug, Error)]
pub enum HealthManagerError {
    #[error("backoff record for unknown account id {0}")]
    UnknownAccount(i64),
    #[error("model-scoped backoff {reason} for account {account_id} has no model")]
    MissingModel {
        account_id: i64,
        reason: BackoffReason,
    },
    #[error("nonterminal backoff {reason} has no expiry")]
    MissingExpiry { reason: BackoffReason },
}

#[derive(Debug, Clone)]
pub struct AccountHealth {
    pub account_id: Option<i64>,
    pub account_name: String,
    pub is_healthy: bool,
    pub health_state: String,
    pub last_check: f64,
    pub last_success: Option<f64>,
    pub last_failure: Option<f64>,
    pub last_failure_category: Option<BackoffReason>,
    pub consecutive_failures: u32,
    pub consecutive_cooldowns: u32,
    pub disabled_until: Option<f64>,
    pub disabled_reason: String,
    pub cooldown_until: f64,
    pub disabled_models: BTreeMap<String, Option<f64>>,
    pub terminal_models: BTreeSet<String>,
    pub circuit_breaker: CircuitBreaker,
}

#[derive(Debug, Clone, PartialEq)]
pub struct AccountHealthSnapshot {
    pub account_id: Option<i64>,
    pub account_name: String,
    pub is_healthy: bool,
    pub health_state: String,
    pub last_check: f64,
    pub last_success: Option<f64>,
    pub last_failure: Option<f64>,
    pub last_failure_category: Option<BackoffReason>,
    pub consecutive_failures: u32,
    pub consecutive_cooldowns: u32,
    pub disabled_until: Option<f64>,
    pub disabled_reason: String,
    pub cooldown_until: f64,
    pub disabled_models: BTreeMap<String, Option<f64>>,
    pub terminal_models: BTreeSet<String>,
    pub circuit: CircuitStats,
}

impl AccountHealth {
    fn snapshot(&self) -> AccountHealthSnapshot {
        AccountHealthSnapshot {
            account_id: self.account_id,
            account_name: self.account_name.clone(),
            is_healthy: self.is_healthy,
            health_state: self.health_state.clone(),
            last_check: self.last_check,
            last_success: self.last_success,
            last_failure: self.last_failure,
            last_failure_category: self.last_failure_category,
            consecutive_failures: self.consecutive_failures,
            consecutive_cooldowns: self.consecutive_cooldowns,
            disabled_until: self.disabled_until,
            disabled_reason: self.disabled_reason.clone(),
            cooldown_until: self.cooldown_until,
            disabled_models: self.disabled_models.clone(),
            terminal_models: self.terminal_models.clone(),
            circuit: self.circuit_breaker.stats(),
        }
    }

    fn is_disabled(&self, now: f64) -> bool {
        self.disabled_until.is_some_and(|until| now < until) || now < self.cooldown_until
    }

    fn is_model_disabled(&self, model_id: &str, now: f64) -> bool {
        if self.is_disabled(now) {
            return true;
        }
        match self.disabled_models.get(model_id) {
            None => false,
            Some(None) => true,
            Some(Some(until)) => now < *until,
        }
    }
}

/// Process-local health manager. Its map lock is synchronous and no lock is
/// held across async work; the breaker has its own short synchronous lock.
#[derive(Clone)]
pub struct HealthManager {
    accounts: Arc<Mutex<BTreeMap<String, AccountHealth>>>,
    clock: Arc<dyn Fn() -> f64 + Send + Sync>,
}

impl std::fmt::Debug for HealthManager {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter
            .debug_struct("HealthManager")
            .field(
                "account_count",
                &self.accounts.lock().expect("health lock").len(),
            )
            .finish()
    }
}

impl Default for HealthManager {
    fn default() -> Self {
        let start = std::time::Instant::now();
        Self::with_clock(move || start.elapsed().as_secs_f64())
    }
}

impl HealthManager {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_clock(clock: impl Fn() -> f64 + Send + Sync + 'static) -> Self {
        Self {
            accounts: Arc::new(Mutex::new(BTreeMap::new())),
            clock: Arc::new(clock),
        }
    }

    pub fn register_account(&self, account_id: i64, account_name: &str) {
        let mut accounts = self.accounts.lock().expect("health lock");
        let clock = Arc::clone(&self.clock);
        accounts
            .entry(account_name.to_owned())
            .or_insert_with(|| AccountHealth {
                account_id: Some(account_id),
                account_name: account_name.to_owned(),
                is_healthy: true,
                health_state: "healthy".to_owned(),
                last_check: self.now(),
                last_success: None,
                last_failure: None,
                last_failure_category: None,
                consecutive_failures: 0,
                consecutive_cooldowns: 0,
                disabled_until: None,
                disabled_reason: String::new(),
                cooldown_until: 0.0,
                disabled_models: BTreeMap::new(),
                terminal_models: BTreeSet::new(),
                circuit_breaker: CircuitBreaker::with_clock(move || clock(), 5, 300.0, 1),
            });
        if let Some(account) = accounts.get_mut(account_name) {
            account.account_id = Some(account_id);
        }
    }

    pub fn accounts(&self) -> Vec<AccountHealthSnapshot> {
        self.accounts
            .lock()
            .expect("health lock")
            .values()
            .map(AccountHealth::snapshot)
            .collect()
    }

    pub fn snapshot(&self, account_name: &str) -> Option<AccountHealthSnapshot> {
        self.accounts
            .lock()
            .expect("health lock")
            .get(account_name)
            .map(AccountHealth::snapshot)
    }

    pub fn get_account_health(&self, account_name: &str) -> Option<AccountHealthSnapshot> {
        self.snapshot(account_name)
    }

    pub fn is_account_healthy(&self, account_name: &str) -> bool {
        self.is_account_healthy_read_only(account_name)
    }

    pub fn is_model_healthy(&self, account_name: &str, model_id: &str) -> bool {
        self.is_model_healthy_read_only(account_name, model_id)
    }

    pub fn get_healthy_accounts(&self, account_names: &[String]) -> Vec<String> {
        account_names
            .iter()
            .filter(|name| self.is_account_healthy(name))
            .cloned()
            .collect()
    }

    pub fn is_account_healthy_read_only(&self, account_name: &str) -> bool {
        let accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get(account_name) else {
            return true;
        };
        let now = self.now();
        self.account_effectively_healthy(account, now) && account.circuit_breaker.can_request()
    }

    pub fn is_model_healthy_read_only(&self, account_name: &str, model_id: &str) -> bool {
        let accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get(account_name) else {
            return true;
        };
        let now = self.now();
        self.account_effectively_healthy(account, now)
            && !account.is_model_disabled(model_id, now)
            && account.circuit_breaker.can_request()
    }

    /// Acquire the breaker probe slot after the caller has completed its
    /// read-only candidate checks. This is the only health claim operation.
    pub fn try_acquire_request(&self, account_name: &str, model_id: &str) -> bool {
        let accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get(account_name) else {
            return false;
        };
        let now = self.now();
        if !self.account_effectively_healthy(account, now)
            || account.is_model_disabled(model_id, now)
        {
            return false;
        }
        account.circuit_breaker.allow_request()
    }

    pub fn release_request(&self, account_name: &str) {
        if let Some(account) = self.accounts.lock().expect("health lock").get(account_name) {
            account.circuit_breaker.release_probe();
        }
    }

    pub fn record_success(&self, account_name: &str, model_id: Option<&str>) {
        let now = self.now();
        let mut accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get_mut(account_name) else {
            return;
        };
        account.consecutive_failures = 0;
        account.consecutive_cooldowns = 0;
        account.last_check = now;
        account.last_success = Some(now);
        if account.disabled_reason.is_empty() && account.health_state != "authentication_failed" {
            account.is_healthy = true;
            account.health_state = "healthy".to_owned();
            account.cooldown_until = 0.0;
        }
        if let Some(model_id) = model_id {
            if !account.terminal_models.contains(model_id) {
                account.disabled_models.remove(model_id);
            }
        }
        account.circuit_breaker.record_success();
    }

    pub fn record_failure(&self, account_name: &str, reason: BackoffReason) {
        let now = self.now();
        let mut accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get_mut(account_name) else {
            return;
        };
        account.consecutive_failures = account.consecutive_failures.saturating_add(1);
        account.last_check = now;
        account.last_failure = Some(now);
        account.last_failure_category = Some(reason);
        account.circuit_breaker.record_failure();
        if reason == BackoffReason::AuthenticationFailed {
            account.health_state = reason.to_string();
            account.is_healthy = false;
            account.disabled_reason = reason.to_string();
            account.disabled_until = None;
        }
    }

    pub fn record_cooldown(&self, account_name: &str, reason: BackoffReason, delay: f64) {
        let now = self.now();
        let mut accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get_mut(account_name) else {
            return;
        };
        let delay = if delay.is_finite() && delay >= 0.0 {
            delay.min(MAX_NONTERMINAL_BACKOFF_SECONDS)
        } else {
            0.0
        };
        account.consecutive_cooldowns = account.consecutive_cooldowns.saturating_add(1);
        account.last_check = now;
        account.last_failure = Some(now);
        account.last_failure_category = Some(reason);
        account.cooldown_until = account.cooldown_until.max(now + delay);
        account.health_state = reason.to_string();
    }

    pub fn disable_account(&self, account_name: &str, reason: &str, duration: Option<f64>) {
        let now = self.now();
        let mut accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get_mut(account_name) else {
            return;
        };
        account.is_healthy = false;
        account.disabled_reason = reason.to_owned();
        account.disabled_until = duration.map(|value| {
            let value = if value.is_finite() {
                value.max(0.0)
            } else {
                0.0
            };
            now + value.min(MAX_NONTERMINAL_BACKOFF_SECONDS)
        });
        account.health_state = reason.to_owned();
    }

    pub fn enable_account(&self, account_name: &str) {
        let now = self.now();
        let mut accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get_mut(account_name) else {
            return;
        };
        account.is_healthy = true;
        account.health_state = "healthy".to_owned();
        account.disabled_until = None;
        account.disabled_reason.clear();
        account.cooldown_until = 0.0;
        account.last_check = now;
        account.circuit_breaker.reset();
    }

    pub fn disable_model(
        &self,
        account_name: &str,
        model_id: &str,
        duration: Option<f64>,
        terminal: bool,
    ) {
        let now = self.now();
        let mut accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get_mut(account_name) else {
            return;
        };
        if terminal || duration.is_none() {
            account.terminal_models.insert(model_id.to_owned());
        } else {
            account.terminal_models.remove(model_id);
        }
        account.disabled_models.insert(
            model_id.to_owned(),
            duration.map(|value| {
                let value = if value.is_finite() {
                    value.max(0.0)
                } else {
                    0.0
                };
                now + value.min(MAX_NONTERMINAL_BACKOFF_SECONDS)
            }),
        );
    }

    pub fn enable_model(&self, account_name: &str, model_id: &str) {
        if let Some(account) = self
            .accounts
            .lock()
            .expect("health lock")
            .get_mut(account_name)
        {
            account.disabled_models.remove(model_id);
            account.terminal_models.remove(model_id);
        }
    }

    pub fn prune_disabled_models(
        &self,
        account_name: &str,
        advertised: &BTreeSet<String>,
    ) -> usize {
        let mut accounts = self.accounts.lock().expect("health lock");
        let Some(account) = accounts.get_mut(account_name) else {
            return 0;
        };
        let stale: Vec<String> = account
            .disabled_models
            .keys()
            .filter(|model| !advertised.contains(*model))
            .cloned()
            .collect();
        for model in &stale {
            account.disabled_models.remove(model);
            account.terminal_models.remove(model);
        }
        stale.len()
    }

    /// Hydrate restart hints using a wall-clock remaining duration and anchor
    /// every resulting deadline in the process-local monotonic domain.
    pub fn hydrate_backoffs(
        &self,
        records: &[AccountBackoffRecord],
        account_names: &BTreeMap<i64, String>,
        wall_now: f64,
    ) -> Result<usize, HealthManagerError> {
        let mut applied = 0;
        for record in records {
            let Some(account_name) = account_names.get(&record.account_id) else {
                return Err(HealthManagerError::UnknownAccount(record.account_id));
            };
            let reason = record.reason;
            if reason == BackoffReason::AuthenticationFailed {
                self.disable_account(account_name, reason.as_str(), None);
                applied += 1;
                continue;
            }
            let Some(deadline) = record.backoff_until_epoch else {
                return Err(HealthManagerError::MissingExpiry { reason });
            };
            let remaining = (deadline - wall_now).max(0.0);
            if remaining == 0.0 {
                continue;
            }
            let remaining = remaining.min(MAX_NONTERMINAL_BACKOFF_SECONDS);
            if reason == BackoffReason::ModelUnavailable {
                let Some(model_id) = record.model_id.as_deref() else {
                    return Err(HealthManagerError::MissingModel {
                        account_id: record.account_id,
                        reason,
                    });
                };
                self.disable_model(account_name, model_id, Some(remaining), false);
            } else {
                self.record_cooldown(account_name, reason, remaining);
            }
            applied += 1;
        }
        Ok(applied)
    }

    fn now(&self) -> f64 {
        (self.clock)()
    }

    fn account_effectively_healthy(&self, account: &AccountHealth, now: f64) -> bool {
        if account.disabled_until.is_some_and(|until| now < until) || now < account.cooldown_until {
            return false;
        }
        let timed_disable_expired = if account.disabled_reason.is_empty() {
            false
        } else if account.health_state == "authentication_failed"
            || account.disabled_until.is_none()
        {
            return false;
        } else {
            true
        };
        timed_disable_expired
            || account.is_healthy
            || matches!(
                account.health_state.as_str(),
                "cooldown" | "quota_exhausted" | "rate_limited" | "operator"
            )
    }
}
